### DISCUSSION TITLE
Has someone implemented fetch?

### DISCUSSION THREAD

**Discussion by SebastianStehle** [sebastienros/jint#1999] [General]
Hi,
I am just curious if someone has implemented ´fetch` already. I have a few http helper functions for my scripts but fetch would give more flexibility.

**Comment by lahma** [comment]
I haven't, but there's https://github.com/AngleSharp/AngleSharp.Js , which doesn't seem to have such support, but might be a logical place to extend and consume such features.

**Comment by sebastienros** [comment]
Since I am ignorant about it I found some docs about it https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API.

Maybe some of these "well established" APIs could be opted-in like in options or with `engine.AddWebFetchApi()`.

Even something like file system apis could be useful (assuming Jint could also provide different kinds of filesystems).

**Comment by CharlieDigital** [comment]
**Edit:** Fixed it with `options => options.ExperimentalFeatures = ExperimentalFeature.TaskInterop` :)

I'm trying to implement a `fetch` compatible API but running into a small problem on the implementation of `Response` analogue.

Specifically, this test fails:

```csharp
[Fact]
public void Can_Handle_Async_Json_Parse_As_JS_From_Object()
{
    var engine = new Engine();

    // We don't need the full functionality here because there is a failure in the async/await we are testing
    var fetcher = new Fetcher();
    engine.SetValue("fetch", fetcher.FetchAsync);

    var messages = new List<string>();
    engine.SetValue(
        "log",
        new Action<object>(msg =>
        {
            messages.Add(msg?.ToString() ?? string.Empty);
        })
    );

    var script = """
        (async () => {
            const response = await fetch();
            log(response);
            const json = await response.json(); // 👈 This await of the Task<JObject> seems to be ignored?
            log(json);  // 👈 This is returns Task<JObject> and not JObject
            return await json.text;
        })()
        """;

    var result = engine.Evaluate(script).UnwrapIfPromise();

    Assert.NotNull(result);
    Assert.Equal("Hello, World!", result);
}

class Fetcher
{
    public async Task<Response> FetchAsync()
    {
        return await Task.FromResult(new Response());
    }
}

class Response
{
    // 👇 This is the cause of the failure
    public async Task<JObject> json()
    {
        return await Task.FromResult(
            JObject.Parse("""{ "text": "Hello, World!" }""")
        );
    }
}
```

Specifically, the line `const json = await response.json();` is returning a `Task<JObject>` and not resolving the Promise/Task.  I can fix this by using `.Result` and making `Response.json()` non-async, but would like to keep it async if possible so it has full fidelity.

Any thoughts on what I'm doing wrong here?  I've looked through the examples of the tests for async and can't quite get this one working.

**Comment by CharlieDigital** [comment]
I ended up implementing a full MCP server for executing LLM generated JavaScript using Jint and gave it a `fetch` analogue :)

https://github.com/CharlieDigital/runjs

**Reply by rocklau** [reply]
cool
